File: go-version/pkg/classifier/classifying.go
package classifier

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"strings"
	"time"

	"github.com/sashabaranov/go-openai"
)

type Config struct {
	OpenAI              OpenAIConfig
	EmailClassification EmailClassificationConfig
}

type OpenAIConfig struct {
	Endpoint    string
	Model       string
	MaxTokens   int
	Temperature int
	APIKey      string
}

type EmailClassificationConfig struct {
	SystemMessage      string
	UserPromptTemplate string
}

type Classifier struct {
	openAIClient *openai.Client
	config       *Config
	ctx          context.Context
}

type llmDecision struct {
	Important   bool   `json:"important"`
	Explanation string `json:"explanation"`
}

func NewClassifier(cfg *Config, ctx context.Context) *Classifier {
	c := openai.DefaultConfig(cfg.OpenAI.APIKey)
	c.BaseURL = normalizeOpenAIBaseURL(cfg.OpenAI.Endpoint)
	client := openai.NewClientWithConfig(c)
	return &Classifier{
		openAIClient: client,
		config:       cfg,
		ctx:          ctx,
	}

}

// normalizeOpenAIBaseURL ensures the BaseURL is suitable for go-openai client
// - appends "/v1" if missing
// - trims any path after "/v1/" if a full endpoint URL was provided
func normalizeOpenAIBaseURL(endpoint string) string {
	if endpoint == "" {
		return ""
	}

	e := strings.TrimRight(endpoint, "/")
	if strings.HasSuffix(e, "/v1") {
		return e
	}

	if idx := strings.Index(e, "/v1/"); idx != -1 {
		return e[:idx+3]
	}

	return e + "/v1"
}

// ClassifyEmail calls OpenAI Chat Completion API to decide if an email is important.
// Returns (important bool, explanation string).
func (c *Classifier) ClassifyEmail(text string) (bool, string) {
	// Build a concise prompt and send to OpenAI-compatible API using SDK
	prompt := fmt.Sprintf(c.config.EmailClassification.UserPromptTemplate, text)

	ctx, cancel := context.WithTimeout(c.ctx, 30*time.Second)
	defer cancel()

	resp, err := c.openAIClient.CreateChatCompletion(ctx, openai.ChatCompletionRequest{
		Model:       c.config.OpenAI.Model,
		MaxTokens:   c.config.OpenAI.MaxTokens,
		Temperature: float32(c.config.OpenAI.Temperature),
		Messages: []openai.ChatCompletionMessage{
			{Role: openai.ChatMessageRoleSystem, Content: c.config.EmailClassification.SystemMessage},
			{Role: openai.ChatMessageRoleUser, Content: prompt},
		},
	})
	if err != nil {
		log.Fatalf("OpenAI request failed: %v", err)
		// return false, "request failed"
	}
	if len(resp.Choices) == 0 {
		log.Printf("No choices in OpenAI response")
		return false, "no choices"
	}

	content := strings.TrimSpace(resp.Choices[0].Message.Content)

	return parseLLMDecision(content)
}

func parseLLMDecision(content string) (bool, string) {
	// Try to extract JSON from the assistant content
	// find first '{' and last '}'
	start := strings.Index(content, "{")
	end := strings.LastIndex(content, "}")
	if start != -1 && end != -1 && end > start {
		jsonStr := content[start : end+1]
		var j llmDecision
		if err := json.Unmarshal([]byte(jsonStr), &j); err == nil {
			return j.Important, j.Explanation
		}
	}

	// Fallback: simple heuristic, check for keywords
	lower := strings.ToLower(content)
	if strings.Contains(lower, "true") || strings.Contains(lower, "important") || strings.Contains(lower, "yes") {
		return true, strings.SplitN(content, "\n", 2)[0]
	}

	return false, strings.SplitN(content, "\n", 2)[0]
}

File: python-version/src/classifier/classifying.py
"""
Email classification using OpenAI API
"""
import json
import re
from typing import Tuple, Optional
from dataclasses import dataclass
import openai
from openai import OpenAI


@dataclass
class OpenAIConfig:
    """OpenAI API configuration"""
    api_key: str
    endpoint: str
    model: str
    max_tokens: int
    temperature: float


@dataclass
class EmailClassificationConfig:
    """Email classification configuration"""
    system_message: str
    user_prompt_template: str


@dataclass
class ClassifierConfig:
    """Classifier configuration"""
    openai: OpenAIConfig
    email_classification: EmailClassificationConfig


class EmailClassifier:
    """Email classifier using OpenAI API"""
    
    def __init__(self, config: ClassifierConfig):
        self.config = config
        
        # Initialize OpenAI client
        client_config = {
            "api_key": config.openai.api_key,
        }
        
        # Set custom endpoint if provided
        if config.openai.endpoint and config.openai.endpoint != "https://api.openai.com/v1":
            client_config["base_url"] = self._normalize_endpoint(config.openai.endpoint)
        
        self.client = OpenAI(**client_config)
    
    def _normalize_endpoint(self, endpoint: str) -> str:
        """
        Normalize OpenAI endpoint URL.
        Ensures the endpoint ends with /v1 if it doesn't already.
        """
        endpoint = endpoint.rstrip('/')
        
        if endpoint.endswith('/v1'):
            return endpoint
        elif '/v1/' in endpoint:
            # Extract base URL up to /v1
            return endpoint[:endpoint.find('/v1') + 3]
        else:
            return endpoint + '/v1'
    
    def classify_email(self, email_text: str) -> Tuple[bool, str]:
        """
        Classify an email as important or unimportant.
        
        Args:
            email_text: The email content to classify
            
        Returns:
            Tuple of (is_important: bool, explanation: str)
        """
        # Build the prompt
        user_prompt = self.config.email_classification.user_prompt_template % email_text
        
        try:
            # Make API call
            response = self.client.chat.completions.create(
                model=self.config.openai.model,
                messages=[
                    {
                        "role": "system",
                        "content": self.config.email_classification.system_message
                    },
                    {
                        "role": "user", 
                        "content": user_prompt
                    }
                ],
                max_tokens=self.config.openai.max_tokens,
                temperature=self.config.openai.temperature
            )
            
            if not response.choices:
                print("No choices in OpenAI response")
                return False, "no choices"
            
            content = response.choices[0].message.content.strip()
            return self._parse_llm_decision(content)
            
        except Exception as e:
            print(f"OpenAI request failed: {e}")
            return False, f"request failed: {e}"
    
    def _parse_llm_decision(self, content: str) -> Tuple[bool, str]:
        """
        Parse the LLM response to extract importance decision and explanation.
        
        Args:
            content: Raw LLM response content
            
        Returns:
            Tuple of (is_important: bool, explanation: str)
        """
        # Try to extract JSON from the response
        json_match = re.search(r'\{[^}]*\}', content)
        if json_match:
            try:
                json_str = json_match.group()
                decision = json.loads(json_str)
                
                if isinstance(decision, dict):
                    important = decision.get('important', False)
                    explanation = decision.get('explanation', '')
                    return important, explanation
            except json.JSONDecodeError:
                pass
        
        # Fallback: simple heuristic based on keywords
        content_lower = content.lower()
        if any(keyword in content_lower for keyword in ['true', 'important', 'yes']):
            # Take first line as explanation
            explanation = content.split('\n')[0] if content else "classified as important"
            return True, explanation
        
        # Default to unimportant
        explanation = content.split('\n')[0] if content else "classified as unimportant"
        return False, explanation
